File: apps/todo_service/src/app/HTTP/Controllers/TodoController.php
<?php

namespace App\HTTP\Controllers;

use App\Domain\DTOs\TodoDTO;
use App\Domain\Services\TodoService;
use App\Lib\HTTPUtils;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class TodoController
{
    private $todoService;

    public function __construct(TodoService $todoService) {
        $this->todoService = $todoService;
    }
    
    public function index(Request $request, Response $response, array $arg)
    {
        $todos = $this->todoService->getAll();
        return HTTPUtils::jsonResponse($response, 201, $todos);
    }

    public function show(Request $request, Response $response, array $args)
    {
        $todo = $this->todoService->get($args['todoId']);
        return HTTPUtils::jsonResponse($response, 200, $todo);;
    }

    public function create(Request $request, Response $response, array $args)
    {
        $payload = $request->getParsedBody();
        $todo = $this->todoService->create(new TodoDTO(
            title: $payload["title"],
            description: $payload["description"],
        ));
        return HTTPUtils::jsonResponse($response, 201, $todo);
    }

    public function update(Request $request, Response $response, array $args)
    {
        $payload = $request->getParsedBody();
        $todo = $this->todoService->update($args['todoId'], new TodoDTO(
            title: $payload["title"],
            description: $payload["description"],
        ));
        return HTTPUtils::jsonResponse($response, 200, $todo);;
    }

    public function destroy(Request $request, Response $response, array $args)
    {
        $this->todoService->destroy($args['todoId']);
        return $response->withStatus(204)->getBody()->write("No content");
    }
}

File: apps/python_service/app/http/controllers/todo_controller.py
from flask import request, Response
from app.domain.services.todo_service import TodoService
from app.domain.dtos.todo_dto import TodoDTO


class TodoController:
    todo_service: TodoService = {}

    def __init__(self, todo_service: TodoService):
        self.todo_service = todo_service

    def index(self) -> Response:
        todos = self.todo_service.get_all()
        return todos, 200

    def show(self, todo_id: int) -> Response:
        todo = self.todo_service.get(todo_id)
        return todo, 200

    def create(self) -> Response:
        payload = request.get_json(force=True)
        todo_dto = TodoDTO(
            title=payload['title'],
            description=payload['description'],
        )
        todo = self.todo_service.create(todo_dto)
        return todo, 201

    def update(self, todo_id: int) -> Response:
        payload = request.get_json()
        todo_dto = TodoDTO(
            title=payload['title'],
            description=payload['description'],
        )
        todo = self.todo_service.update(todo_id, todo_dto)
        return todo, 200

    def destroy(self, todo_id: int) -> Response:
        self.todo_service.destroy(todo_id)
        return "No content", 204
